src/Action.hs

```diff
@@ -91,47 +102,61 @@ makeLenses ''UserState
 userLoggedOut :: UserState
 userLoggedOut = UserState Nothing Nothing Nothing
 
-data ActionEnv r = ActionEnv
-    { _persistNat :: r :~> ExceptT PersistExcept IO
-    , _config     :: Config
+data ActionEnv = ActionEnv
+    { _envRunPersist :: RunPersist
+    , _envConfig     :: Config
     }
 
 makeLenses ''ActionEnv
 
-instance GetCsrfSecret (ActionEnv r) where
-    csrfSecret = config . csrfSecret
+instance GetCsrfSecret ActionEnv where
+    csrfSecret = envConfig . csrfSecret
 
 -- | Top level errors can happen.
 --
--- FIXME: this will have a constructor dedicated for PersistExcept, and 'ServantErr' will only be
--- introduced later.
-newtype ActionExcept = ActionExcept { unActionExcept :: ServantErr }
+-- FIXME: 'ServantErr' should be abstracted away.
+data ActionExcept
+    = ActionExcept { unActionExcept :: ServantErr }
+    | ActionPersistExcept PersistExcept
     deriving (Eq, Show)
 
 makePrisms ''ActionExcept
 
-class ( ActionLog m
-      , ActionPersist r m
+type ActionM m =
+      ( ActionLog m
+      , ActionPersist m
       , ActionUserHandler m
       , ActionError m
       , ActionTempCsvFiles m
-      ) => ActionM r m
+      , ActionRandomPassword m
+      , ActionCurrentTimestamp m
+      )
 
 class Monad m => ActionLog m where
     -- | Log events
     logEvent :: ST -> m ()
 
--- | A monad that can include actions changing a persistent state.
---
--- @r@ is determined by @m@, because @m@ is intended to be the program's
--- action monad, so @r@ is just the persistent implementation chosen
--- to be used in the action monad.
-class (PersistM r, Monad m, MonadError ActionExcept m) => ActionPersist r m | m -> r where
-    -- | Run @Persist@ computation in the action monad.
-    -- Authorization of the action should happen here.
-    -- FIXME: Rename atomically, and only call on
-    -- complex computations.
-    persistent :: r a -> m a
+-- | A monad that can run acid-state.
+class (MonadError ActionExcept m) => ActionPersist m where
+    aqueryDb :: m AulaData
```

Comment:
I feel like the `a` prefix in this class is a leftover from the beginning of the refactor. In my opinion `queryDb`, `query`, `mquery`, `equery`.